plugins/outputs/opentelemetry/client.go

```diff
@@ -0,0 +1,225 @@
+func (c *client) ping(ctx context.Context) error {
+	// Loop until the context is canceled, allowing for retryable failures.
+	for {
+		err := c.connect(ctx)
+
+		if err == nil {
+			service := metricsService.NewMetricsServiceClient(c.conn)
+			empty := &metricsService.ExportMetricsServiceRequest{}
+
+			_, err = service.Export(metadata.NewOutgoingContext(ctx, c.headers), empty)
+			if err == nil {
+				return nil
+			}
+		}
+
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		default:
+			if isRecoverable(err) {
+				c.logger.Infof("ping recoverable error, still trying, err=%w", err)
+				continue
+			}
+		}
+		return fmt.Errorf("non-recoverable failure in ping, err=%w", err)
+	}
+}
+
+// store sends a batch of samples to the endpoint.
+func (c *client) store(samples []*metricspb.ResourceMetrics) error {
+	metricsLen := len(samples)
+	if metricsLen == 0 {
+		// Nothing to do, return silently.
+		return nil
+	}
+	req := metricsService.ExportMetricsServiceRequest{
+		ResourceMetrics: samples,
+	}
+
+	// Note the call to connect() applies its own timeout for Dial().
+	ctx := context.Background()
+	err := c.connect(ctx)
+	if err != nil {
+		return err
+	}
+
+	ctx, cancel := context.WithTimeout(ctx, c.timeout)
+	defer cancel()
+
+	service := metricsService.NewMetricsServiceClient(c.conn)
+
+	errs := make(chan error, metricsLen/maxTimeseriesPerRequest+1)
+	var wg sync.WaitGroup
+	for i := 0; i < metricsLen; i += maxTimeseriesPerRequest {
+		end := i + maxTimeseriesPerRequest
+		if end > metricsLen {
+			end = metricsLen
+		}
+		wg.Add(1)
+		go func(begin int, end int) {
+			defer wg.Done()
+			reqCopy := &metricsService.ExportMetricsServiceRequest{
+				ResourceMetrics: req.ResourceMetrics[begin:end],
+			}
+
+			var md metadata.MD
+			var err error
+
+			if _, err = service.Export(metadata.NewOutgoingContext(ctx, c.headers), reqCopy, grpc.Trailer(&md)); err != nil {
+				c.logger.Errorf("export failure, err=%w size=%d trailers=%v recoverable=%t",
+					err,
+					proto.Size(reqCopy),
+					md,
+					isRecoverable(err),
+				)
+				errs <- err
+				return
+			}
+
+			c.logger.Debug("successful write, records=%d size=%d trailers=%v", end-begin, proto.Size(reqCopy), md)
+		}(i, end)
+	}
+	wg.Wait()
+	close(errs)
+	if err, ok := <-errs; ok {
+		return err
+	}
+	return nil
+}
+
+func (c *client) close() error {
+	if c.conn == nil {
+		return nil
+	}
+	return c.conn.Close()
+}
+
+func isRecoverable(err error) bool {
+	if errors.Is(err, context.Canceled) {
+		return true
+	}
+
+	if errors.Is(err, context.DeadlineExceeded) {
+		return true
+	}
+
+	s, ok := status.FromError(err)
+	if !ok {
+		return false
+	}
+	switch s.Code() {
+	case codes.DeadlineExceeded, codes.Canceled, codes.ResourceExhausted,
+		codes.Aborted, codes.OutOfRange, codes.Unavailable, codes.DataLoss:
+		// See https://github.com/open-telemetry/opentelemetry-specification/
+		// blob/master/specification/protocol/otlp.md#response
+		return true
+	default:
+		return false
+	}
```

Comment:
You might change this to 
```suggestion
	}
	return false
```
but it's a matter of taste...